File: shared/scripts/test_pick_destination.py
#!/usr/bin/env python3
"""Offline tests for the shared Python destination picker."""
import importlib.util
import io
import json
import pathlib
import unittest
from contextlib import redirect_stdout


SCRIPT = pathlib.Path(__file__).with_name("pick_destination.py")
SPEC = importlib.util.spec_from_file_location("pick_destination", SCRIPT)
picker = importlib.util.module_from_spec(SPEC)
SPEC.loader.exec_module(picker)


class DestinationPickerTest(unittest.TestCase):
    def setUp(self):
        self.original_call = picker.call
        self.routes = {}
        self.calls = []
        picker.call = self.fake_call

    def tearDown(self):
        picker.call = self.original_call

    def fake_call(self, method, path, body=None):
        self.calls.append((method, path, body))
        return self.routes.get((method, path), {})

    def route(self, method, path, response):
        self.routes[(method, path)] = response

    def test_my_documents_uses_member_home_folder_id(self):
        self.route("GET", "/v2/whoami", {"userId": "member-1"})
        self.route(
            "GET",
            "/v2/members/member-1",
            {"homeFolderId": "home-folder-id"},
        )

        self.assertEqual("home-folder-id", picker.my_documents_id())
        self.assertEqual(
            [
                ("GET", "/v2/whoami", None),
                ("GET", "/v2/members/member-1", None),
            ],
            self.calls,
        )

    def test_legacy_fallback_uses_folder_id_not_parent_id(self):
        self.route("GET", "/v2/whoami", {"userId": "member-1"})
        self.route("GET", "/v2/members/member-1", {})
        self.route(
            "GET",
            "/v2/members/member-1/files?typeFilters=folder&limit=500",
            {
                "entries": [
                    {
                        "id": "my-documents-id",
                        "parentId": "wrong-parent-id",
                        "name": "My Documents",
                    }
                ]
            },
        )

        self.assertEqual("my-documents-id", picker.my_documents_id())

    def test_list_exposes_authoritative_my_documents_id(self):
        self.route("GET", "/v2/whoami", {"userId": "member-1"})
        self.route(
            "GET",
            "/v2/members/member-1",
            {"homeFolderId": "home-folder-id"},
        )
        self.route(
            "GET",
            "/v2/workspaces?limit=500",
            {"entries": [{"workspaceId": "workspace-1", "name": "Sales"}]},
        )
        self.route(
            "GET",
            "/v2/files?typeFilters=folder&limit=500",
            {
                "entries": [
                    {
                        "id": "folder-1",
                        "name": "Reports",
                        "parentId": "workspace-1",
                        "permission": "edit",
                    }
                ]
            },
        )
        output = io.StringIO()

        with redirect_stdout(output):
            picker.cmd_list()

        result = json.loads(output.getvalue())
        self.assertEqual("home-folder-id", result["myDocuments"])
        self.assertEqual("Sales", result["folders"][0]["parentName"])

    def test_create_defaults_parent_to_home_folder(self):
        self.route("GET", "/v2/whoami", {"userId": "member-1"})
        self.route(
            "GET",
            "/v2/members/member-1",
            {"homeFolderId": "home-folder-id"},
        )
        self.route(
            "POST",
            "/v2/files",
            {
                "id": "new-folder",
                "name": "Migrated",
                "parentId": "home-folder-id",
            },
        )
        output = io.StringIO()

        with redirect_stdout(output):
            picker.cmd_create(["--name", "Migrated"])

        post = self.calls[-1]
        self.assertEqual(
            {"type": "folder", "name": "Migrated", "parentId": "home-folder-id"},
            post[2],
        )


if __name__ == "__main__":
    unittest.main(verbosity=2)

File: plugins/cognos-to-sigma/skills/cognos-to-sigma/scripts/test-pick-destination.mjs
#!/usr/bin/env node
// Offline behavioral test for the My Documents resolver.
import assert from 'node:assert/strict';
import fs from 'node:fs';

const source = fs.readFileSync(new URL('./pick-destination.mjs', import.meta.url), 'utf8');
const match = source.match(/async function myDocumentsId\(\) \{[\s\S]*?\n\}/);
assert.ok(match, 'myDocumentsId function is present');

function resolverWith(call) {
  return new Function('call', `${match[0]}; return myDocumentsId;`)(call);
}

{
  const calls = [];
  const responses = {
    '/v2/whoami': { userId: 'member-1' },
    '/v2/members/member-1': { homeFolderId: 'home-folder-id' },
  };
  const resolver = resolverWith(async (method, path) => {
    calls.push([method, path]);
    return responses[path];
  });

  assert.equal(await resolver(), 'home-folder-id');
  assert.deepEqual(calls, [
    ['GET', '/v2/whoami'],
    ['GET', '/v2/members/member-1'],
  ]);
}

{
  const responses = {
    '/v2/whoami': { userId: 'member-1' },
    '/v2/members/member-1': {},
    '/v2/members/member-1/files?typeFilters=folder&limit=500': {
      entries: [{
        id: 'my-documents-id',
        parentId: 'wrong-parent-id',
        path: 'My Documents',
      }],
    },
  };
  const resolver = resolverWith(async (_method, path) => responses[path]);

  assert.equal(await resolver(), 'my-documents-id');
}

const orchestratorSource = fs.readFileSync(
  new URL('./migrate-cognos.mjs', import.meta.url), 'utf8');
const orchestratorMatch = orchestratorSource.match(
  /async function resolveFolder\(\) \{[\s\S]*?\n\}/);
assert.ok(orchestratorMatch, 'orchestrator resolveFolder function is present');

function orchestratorResolverWith(api) {
  const die = (message) => { throw new Error(message); };
  const line = () => {};
  return new Function(
    'api', 'die', 'line', `${orchestratorMatch[0]}; return resolveFolder;`
  )(api, die, line);
}

{
  const calls = [];
  const resolver = orchestratorResolverWith(async (method, path) => {
    calls.push([method, path]);
    if (path === '/v2/whoami') return { status: 200, json: { userId: 'member-1' } };
    if (path === '/v2/members/member-1') {
      return { status: 200, json: { homeFolderId: 'home-folder-id' } };
    }
    throw new Error(`unexpected API call: ${path}`);
  });

  assert.equal(await resolver(), 'home-folder-id');
  assert.deepEqual(calls, [
    ['GET', '/v2/whoami'],
    ['GET', '/v2/members/member-1'],
  ]);
}

{
  const resolver = orchestratorResolverWith(async (_method, path) => {
    if (path === '/v2/whoami') return { status: 200, json: { userId: 'member-1' } };
    if (path === '/v2/members/member-1') return { status: 200, json: {} };
    if (path === '/v2/members/member-1/files') {
      return {
        status: 200,
        json: {
          entries: [{
            id: 'my-documents-id',
            parentId: 'wrong-parent-id',
            path: 'My Documents',
          }],
        },
      };
    }
    throw new Error(`unexpected API call: ${path}`);
  });

  assert.equal(await resolver(), 'my-documents-id');
}

console.log('pick-destination My Documents resolver: PASS');
